File: backend/analysis/flicker.py
"""Voltage flicker assessment (IEC 61000-3-3 / IEC 61000-4-15) — a planning-
level screening study for repetitive voltage changes caused by switching a
fluctuating load, chiefly a motor that starts repeatedly (an intermittent
compressor, pump, or process drive rather than a once-off start).

**Scope and honesty about the method.** A true IEC 61000-4-15 flickermeter
output (Pst/Plt) is derived from the actual AC voltage waveform sampled over
a 10-minute (Pst) / 2-hour (Plt) window through a specific demodulation +
perception-weighting filter chain. A steady-state single-line-diagram tool
has no such waveform to sample, so this study cannot reproduce a certified
flickermeter measurement. What it CAN do rigorously is compute the physical
input every flicker assessment starts from — the **relative voltage change
d(%)** a switching event causes — using the same Thevenin-superposition
machinery the static motor-starting study is built on (`motor_starting.py`),
and then translate that into a **planning-level Pst/Plt estimate** via the
documented simplified method IEC 61000-3-3 itself provides for exactly this
purpose (Annex B): repetitive rectangular voltage steps of magnitude d(%),
occurring r times per minute, have an empirically-curve-fitted severity.

The curve-fit used here is anchored at the single most consistently-cited
reference point in the flicker literature — a ~3 % step at 1 change/minute
corresponds to Pst ≈ 1 (the IEC "Pst = 1" borderline-of-irritability curve) —
with the well-established high-frequency roll-off exponent 0.31:

    Pst_estimate = (d / d_anchor) · r^exponent,   d_anchor = 3.0 (%), exponent = 0.31

Both `d_anchor` and `exponent` are exposed as request parameters so a user
with the actual IEC 61000-3-3 table at hand can recalibrate the estimate.
**This is a screening estimate, not a certified measurement** — a borderline
or failing result should be confirmed against the standard's own curve/table
or by field measurement before being used as a compliance determination.

For Plt, IEC 61000-3-3 permits Plt ≈ Pst when the disturbance source's
emission does not vary materially over the 2-hour assessment window (the
normal case for a single repetitively-switched load) — used here directly.

Compliance limits default to the IEC 61000-3-3 LV connection values
(Pst ≤ 1.0, Plt ≤ 0.65); MV/HV connections are assessed against
utility/IEC 61000-3-7-allocated planning levels, which are project-specific
and passed in as overrides rather than hard-coded.

Results are on-demand (not persisted).
"""

from __future__ import annotations

import math

from ..models.schemas import ProjectData
from .loadflow import run_load_flow, insert_implicit_load_buses
from .motor_starting import (
    _build_adjacency, _find_motor_bus, _thevenin_z1, _solve_pq_dip,
    _STARTING_METHODS,
)

DEFAULT_PST_LIMIT = 1.0
DEFAULT_PLT_LIMIT = 0.65
DEFAULT_D_ANCHOR_PCT = 3.0
DEFAULT_EXPONENT = 0.31


def _starts_per_hour(comp):
    try:
        v = float(comp.props.get("flicker_starts_per_hour", 0) or 0)
        return max(0.0, v)
    except (TypeError, ValueError):
        return 0.0


def _pst_estimate(d_pct, starts_per_hour, d_anchor_pct=DEFAULT_D_ANCHOR_PCT,
                  exponent=DEFAULT_EXPONENT):
    """Planning-level Pst estimate for a repetitive rectangular voltage step
    of magnitude d_pct (%) occurring starts_per_hour times per hour — see the
    module docstring for the method, calibration anchor and its limits.
    Returns 0.0 for a non-repetitive (starts_per_hour <= 0) event."""
    if starts_per_hour <= 0 or d_pct <= 0:
        return 0.0
    r_per_min = starts_per_hour / 60.0
    return (d_pct / d_anchor_pct) * (r_per_min ** exponent)


def run_flicker_analysis(project: ProjectData, pst_limit: float = None,
                         plt_limit: float = None, d_anchor_pct: float = None,
                         exponent: float = None) -> dict:
    pst_limit = DEFAULT_PST_LIMIT if pst_limit is None else float(pst_limit)
    plt_limit = DEFAULT_PLT_LIMIT if plt_limit is None else float(plt_limit)
    d_anchor = DEFAULT_D_ANCHOR_PCT if d_anchor_pct is None else max(0.01, float(d_anchor_pct))
    exp = DEFAULT_EXPONENT if exponent is None else float(exponent)

    project = insert_implicit_load_buses(project)
    comp_map = {c.id: c for c in project.components}
    adj = _build_adjacency(project)

    motors = [c for c in project.components
             if c.type in ("motor_induction", "motor_synchronous")
             and _starts_per_hour(c) > 0]
    if not motors:
        return {"converged": False, "note": (
            "No motors flagged with a repetitive starting rate — set "
            "'Starts per Hour' on any motor that starts intermittently "
            "(a compressor, pump, or process drive) to include it in the "
            "flicker screening. A motor that starts once and runs "
            "continuously does not cause flicker."),
            "warnings": [], "sources": []}

    baseline = run_load_flow(project, "newton_raphson", include_synthetic=True)
    if not baseline.converged:
        baseline = run_load_flow(project, "gauss_seidel", include_synthetic=True)
    if not baseline.converged:
        return {"converged": False,
                "note": "Baseline load flow did not converge.",
                "warnings": [], "sources": []}
    v_pre = {bid: b.voltage_pu for bid, b in (baseline.buses or {}).items()}

    warnings = []
    sources = []
    for motor in motors:
        mp = motor.props
        is_sync = motor.type == "motor_synchronous"
        name = str(mp.get("name", motor.id))
        voltage_kv = float(mp.get("voltage_kv", 0) or 0)
        power_factor = float(mp.get("power_factor", 0.9 if is_sync else 0.85) or 0.85)
        lrc = float(mp.get("locked_rotor_current", 5.5 if is_sync else 6.0) or 6.0)
        starts_hr = _starts_per_hour(motor)

        if is_sync:
            rated_kva = float(mp.get("rated_kva", 0) or 0)
            if rated_kva <= 0 or voltage_kv <= 0:
                warnings.append(f"Motor '{name}' has invalid ratings, skipped.")
                continue
            flc_a = rated_kva / (math.sqrt(3) * voltage_kv)
        else:
            rated_kw = float(mp.get("rated_kw", 0) or 0)
            efficiency = float(mp.get("efficiency", 0.93) or 0.93)
            if rated_kw <= 0 or voltage_kv <= 0:
                warnings.append(f"Motor '{name}' has invalid ratings, skipped.")
                continue
            flc_a = rated_kw / (math.sqrt(3) * voltage_kv * efficiency * power_factor)

        method_key = str(mp.get("starting_method", "dol")).lower()
        factor, method_label = _STARTING_METHODS.get(method_key, _STARTING_METHODS["dol"])
        start_current_a = flc_a if factor is None else flc_a * lrc * factor
        s_start_mva = voltage_kv * start_current_a * math.sqrt(3) / 1000.0

        terminal_bus = _find_motor_bus(motor.id, adj, comp_map)
        if terminal_bus is None:
            warnings.append(f"Motor '{name}': no terminal bus found, skipped.")
            continue
        bus_comp = comp_map.get(terminal_bus)
        bus_name = str(bus_comp.props.get("name", terminal_bus)) if bus_comp else terminal_bus

        z_th = _thevenin_z1(project, terminal_bus, motor.id)
        if z_th is None:
            warnings.append(f"Motor '{name}': no source path found for the "
                            "Thevenin voltage-step calculation, skipped.")
            continue

        v_pre_term = v_pre.get(terminal_bus, 1.0)
        s_pu = s_start_mva / project.baseMVA
        start_pf = 0.3
        s_cplx = s_pu * complex(start_pf, math.sqrt(1 - start_pf ** 2))
        v_start = _solve_pq_dip(v_pre_term, z_th, s_cplx)
        if v_start is None:
            warnings.append(f"Motor '{name}': starting load exceeds the "
                            "network's transfer capability (voltage "
                            "collapse) — flicker screening not meaningful "
                            "until the starting condition itself is fixed.")
            continue

        d_pct = max(0.0, (v_pre_term - v_start) / v_pre_term * 100.0) if v_pre_term > 0 else 0.0
        pst = _pst_estimate(d_pct, starts_hr, d_anchor, exp)
        plt = pst   # stationary repetitive source — IEC 61000-3-3 simplification

        pst_ok = pst <= pst_limit + 1e-9
        plt_ok = plt <= plt_limit + 1e-9
        sources.append({
            "motor_id": motor.id,
            "motor_name": name,
            "terminal_bus": bus_name,
            "starting_method": method_label,
            "starts_per_hour": round(starts_hr, 3),
            "relative_voltage_change_pct": round(d_pct, 3),
            "pst": round(pst, 3),
            "plt": round(plt, 3),
            "pst_limit": pst_limit,
            "plt_limit": plt_limit,
            "pst_compliant": pst_ok,
            "plt_compliant": plt_ok,
            "compliant": bool(pst_ok and plt_ok),
        })

    sources.sort(key=lambda s: -s["pst"])
    overall_compliant = all(s["compliant"] for s in sources) if sources else True

    return {
        "converged": True,
        "sources": sources,
        "compliant": overall_compliant,
        "d_anchor_pct": d_anchor,
        "exponent": exp,
        "method": ("Planning-level screening estimate: relative voltage "
                   "change d(%) from Thevenin-superposition (motor-starting "
                   "machinery), translated to Pst via the IEC 61000-3-3-style "
                   "simplified curve Pst = (d/d_anchor)*r^exponent "
                   f"(default anchor {DEFAULT_D_ANCHOR_PCT}% at 1 change/min, "
                   f"exponent {DEFAULT_EXPONENT}); Plt = Pst for a stationary "
                   "repetitive source. NOT a certified IEC 61000-4-15 "
                   "flickermeter measurement — verify borderline/failing "
                   "results against the standard's own curve or by field "
                   "measurement."),
        "warnings": warnings,
        "note": "",
    }

File: frontend/js/flicker.js
/* ProtectionPro — Voltage Flicker (IEC 61000-3-3 / IEC 61000-4-15) UI.
 *
 * Setup modal (Pst/Plt limits, curve calibration) → backend /analysis/flicker
 * → results modal listing each repetitively-starting motor's relative
 * voltage change, Pst/Plt estimate and IEC 61000-3-3 compliance verdict.
 *
 * Screens motors flagged with a nonzero "Starts per Hour" (Voltage Flicker
 * property section) — a once-off start does not cause flicker. This is a
 * PLANNING-LEVEL SCREENING ESTIMATE (not a certified IEC 61000-4-15
 * flickermeter measurement) — see the results-modal method note.
 *
 * Results are on-demand (not persisted) — re-run after edits.
 */
const Flicker = {
  _result: null,
  _cfg: { pst_limit: 1.0, plt_limit: 0.65, d_anchor_pct: 3.0, exponent: 0.31 },

  _esc(s) {
    return String(s ?? '').replace(/&/g, '&amp;').replace(/</g, '&lt;')
      .replace(/>/g, '&gt;').replace(/"/g, '&quot;');
  },

  openConfig() {
    const c = this._cfg;
    const body = document.getElementById('flk-config-body');
    body.innerHTML = `
      <p style="font-size:12px;color:var(--text-muted,#6d6d6d);margin:0 0 12px">
        Screens every motor with a non-zero <strong>Starts per Hour</strong> (Voltage
        Flicker property section — a once-off start is excluded) for repetitive-switching
        voltage flicker. Relative voltage change is computed by the same Thevenin
        superposition as the Motor Starting study; Pst/Plt are a
        <strong>planning-level estimate</strong> from the IEC 61000-3-3-style simplified
        curve, not a certified IEC 61000-4-15 flickermeter measurement — verify a
        borderline/failing result against the standard's own curve or by field measurement.</p>
      <div style="display:grid;grid-template-columns:auto 1fr;gap:8px 12px;align-items:center;font-size:13px">
        <label for="flk-pst">Pst limit</label>
        <input id="flk-pst" type="number" min="0.1" step="0.05" value="${c.pst_limit}">
        <label for="flk-plt">Plt limit</label>
        <input id="flk-plt" type="number" min="0.1" step="0.05" value="${c.plt_limit}">
        <label for="flk-anchor">Curve anchor d (%) @ 1/min</label>
        <input id="flk-anchor" type="number" min="0.1" step="0.1" value="${c.d_anchor_pct}">
        <label for="flk-exp">Curve exponent</label>
        <input id="flk-exp" type="number" min="0.05" max="1" step="0.01" value="${c.exponent}">
      </div>
      <p style="font-size:11px;color:var(--text-muted,#6d6d6d);margin:12px 0 0">
        Defaults: Pst ≤ 1.0 / Plt ≤ 0.65 are the IEC 61000-3-3 LV connection limits — an
        MV/HV connection typically uses a utility-allocated IEC 61000-3-7 planning level
        instead. The curve anchor/exponent default to the standard's own most-cited
        reference point (≈3% step at 1 change/min ⇒ Pst≈1) and 0.31 roll-off exponent;
        adjust if you have the standard's exact curve to hand.</p>`;
    document.getElementById('flk-config-modal').style.display = '';
  },

  _readConfig() {
    const v = id => document.getElementById(id);
    this._cfg = {
      pst_limit: parseFloat(v('flk-pst').value) || 1.0,
      plt_limit: parseFloat(v('flk-plt').value) || 0.65,
      d_anchor_pct: parseFloat(v('flk-anchor').value) || 3.0,
      exponent: parseFloat(v('flk-exp').value) || 0.31,
    };
    return this._cfg;
  },

  async runConfigured() {
    const c = this._readConfig();
    document.getElementById('flk-config-modal').style.display = 'none';
    const label = 'Running voltage flicker screening…';
    document.getElementById('status-info').textContent = label;
    if (typeof UI !== 'undefined' && UI.setBusy) UI.setBusy(true, label);
    try {
      const result = await API.runFlickerAnalysis({
        pstLimit: c.pst_limit, pltLimit: c.plt_limit,
        dAnchorPct: c.d_anchor_pct, exponent: c.exponent,
      });
      this._result = result;
      this.show(result);
      document.getElementById('status-info').textContent = result.converged
        ? (result.compliant
           ? `Voltage flicker: ${result.sources.length} source(s) screened, all compliant`
           : `Voltage flicker: ${result.sources.filter(s => !s.compliant).length} of ${result.sources.length} source(s) exceed the limit`)
        : 'Voltage flicker screening did not run.';
    } catch (e) {
      console.error('Flicker analysis error:', e);
      document.getElementById('status-info').textContent = 'Voltage flicker screening failed.';
      if (typeof showValidationModal === 'function') {
        showValidationModal('Voltage Flicker — Error', [{ msg: e.message || 'Unknown error' }], [], null);
      }
    } finally {
      if (typeof UI !== 'undefined' && UI.setBusy) UI.setBusy(false);
    }
  },

  show(result) {
    this._result = result;
    const modal = document.getElementById('flk-modal');
    const body = document.getElementById('flk-body');
    if (!modal || !body) return;
    this._render(body);
    modal.style.display = '';
  },

  _render(body) {
    const r = this._result || {};
    let html = '';
    if ((r.warnings || []).length) {
      html += '<div class="af-warnings">' + r.warnings.map(w =>
        `<div class="af-warning-item">⚠ ${this._esc(w)}</div>`).join('') + '</div>';
    }
    if (!r.converged) {
      body.innerHTML = html + `<p style="color:#c62828"><strong>Study did not run.</strong> ${this._esc(r.note || '')}</p>`;
      return;
    }
    const col = r.compliant ? '#2e7d32' : '#c62828';
    const verdict = r.compliant
      ? `All ${r.sources.length} screened source(s) within limit`
      : `${r.sources.filter(s => !s.compliant).length} of ${r.sources.length} source(s) exceed the Pst/Plt limit`;
    html += `<div style="margin-bottom:10px;padding:10px 14px;border-radius:6px;border:1px solid ${col};background:${col}14">
      <span style="font-weight:700;color:${col}">${this._esc(verdict)}</span>
    </div>`;
    html += `<div style="font-size:11px;color:var(--text-muted,#6d6d6d);margin-bottom:12px">${this._esc(r.method || '')}</div>`;

    const rows = (r.sources || []).map(s => {
      const okCol = s.compliant ? '#2e7d32' : '#c62828';
      return `<tr>
        <td>${this._esc(s.motor_name)}</td>
        <td>${this._esc(s.terminal_bus)}</td>
        <td>${this._esc(s.starting_method)}</td>
        <td>${s.starts_per_hour}</td>
        <td>${s.relative_voltage_change_pct}</td>
        <td>${s.pst}${s.pst_compliant ? '' : ' ✗'}</td>
        <td>${s.plt}${s.plt_compliant ? '' : ' ✗'}</td>
        <td style="color:${okCol};font-weight:600">${s.compliant ? 'PASS' : 'FAIL'}</td>
      </tr>`;
    }).join('');
    html += `<table class="af-table" style="font-size:11px;font-variant-numeric:tabular-nums">
      <thead><tr><th>Motor</th><th>Bus</th><th>Starting</th><th>Starts/h</th><th>d (%)</th><th>Pst</th><th>Plt</th><th>Verdict</th></tr></thead>
      <tbody>${rows}</tbody></table>`;
    body.innerHTML = html;
  },
};
